### PR TITLE
Add `messages.json` and update description.

### LINKED ISSUE #80: Update `Messages`.
* Remove `app_ack_proposal`
* Rename `mls_plaintext_application` -> `public_message_application`
* Rename `mls_plaintext_proposal` -> `public_message_proposal`
* Rename `mls_plaintext_commit -> `public_message_commit`
* Rename `mls_ciphertext -> `private_message_ciphertext`

* Remove `capabilities` (as they are part of the leaf node of the key package)
* Remove `lifetime` (as it is part of the leaf node key package)

Is there anything else we want to do directly? If not, I could make a PR in the next days.

* Add `group_context_extensions_proposal`
* Add ciphersuite parameter?

### PR DESCRIPTION
**Pull request by duesee**
A first try on generating test vectors. I'm not sure if I did everything correctly, so please be careful. I guess we will see what happens. This closes #80.

### CODE REVIEW

**Review by bifurcation — COMMENTED** [review]
This looks good to me. The only change I would suggest is to just have one entry in the test vector array -- since this test vector is only testing serialization/deserialization, it doesn't need to depend on the ciphersuite.

(Alternatively, if you really want to test that serialization/deserialization works regardless of ciphersuite, you need to add a `cipher_suite` field!  But as above, I wouldn't go this route.)

File: test-vectors.md [hunk 1]
```diff
@@ -485,30 +485,26 @@ Format:
...
   "commit": /* serialized Commit */,
 
-  "mls_plaintext_application": /* serialized MLSPlaintext(ApplicationData) */,
-  "mls_plaintext_proposal": /* serialized MLSPlaintext(Proposal(*)) */,
-  "mls_plaintext_commit": /* serialized MLSPlaintext(Commit) */,
-  "mls_ciphertext": /* serialized MLSCiphertext */,
+  "public_message_application": /* serialized PublicMessage(ApplicationData) */,
+  "public_message_proposal": /* serialized PublicMessage(Proposal(*)) */,
+  "public_message_commit": /* serialized PublicMessage(Commit) */,
+  "private_message_ciphertext": /* serialized PrivateMessage */,
```
Comment (bifurcation) [inline test-vectors.md:507]:
```suggestion
  "private_message": /* serialized PrivateMessage */,
```

(Including `_ciphertext` might suggest `PrivateMessage.ciphertext`)

**Review by bifurcation — COMMENTED** [review]
One other thing that occurs to me: It might be good to cover all of the `MLSMessage.wire_format` cases.  Right now, we're missing:

* `key_package`
* `group_info`
* `welcome`

A simple way to do that would be to change the fields here that already have that type, so that they provide the value wrapped as an MLSMessage instead of as a raw value.  We could also have both raw and wrapped fields, but that seems duplicative.

**Comment by bifurcation** [comment]
Actually, scratch the above, it doesn't look like we're covering MLSMessage at all.  So yeah, would suggest providing all the types that can be wrapped as MLSMessage in their wrapped form.

**Comment by bifurcation** [comment]
Noticed while attempting to validate `messages.json`:
* `ProtocolVersion` is still one byte
* "OpenMLS rocks", cute :)
* `Capabilities.proposals` only needs to list non-default proposals
* `external_init_proposal` serializes as `0000`, when it should only be `00`.  But it would be nice to have some content here.

With those nits and the above fixed, I've got this passing in a [branch of MLSpp](https://github.com/cisco/mlspp/pull/306), so once they're fixed, we should be able to interop in fairly short order.

**Comment by duesee** [comment]
Thanks for the review! I pushed [9190fee](https://github.com/mlswg/mls-implementations/pull/82/commits/9190fee66421ed72fd1cdb572e7a3c1a8e2fcd54) with the following changes:

> ProtocolVersion is still one byte

Missed a commit in OpenMLS, it's two bytes now.

> "OpenMLS rocks", cute :)

Hehe, I figured I should make it a bit more generic :-)

> Capabilities.proposals only needs to list non-default proposals

Capabilities.proposals only includes AppAck now.

> external_init_proposal serializes as 0000, when it should only be 00. But it would be nice to have some content here.

My fault. I overlooked an old TODO in the tests. It's now an ExternalInit with kem_output == random_bytes(16)

> Actually, scratch the above, it doesn't look like we're covering MLSMessage at all. So yeah, would suggest providing all the types that can be wrapped as MLSMessage in their wrapped form.

Good idea! I applied this idea to all {Public,Private}Message fields as well and extended the comments.

> The only change I would suggest is to just have one entry in the test vector array -- since this test vector is only testing serialization/deserialization, it doesn't need to depend on the ciphersuite.

Could you elaborate on this one? I'm not sure what test vector array you mean?

**Comment by bifurcation** [comment]
@duesee if you fix the WireFormat size issue, I have confirmed that MLSpp passes the current file except for that bug, so we should be clear to merge.

**Comment by duesee** [comment]
Okay, all should be good now.

**Comment by bifurcation** [comment]
Confirmed, this now works with MLSpp

**Pull request merged**